helix-core/src/main/java/org/apache/helix/controller/GenericHelixController.java

```diff
@@ -187,6 +188,16 @@
 
   private HelixManager _helixManager;
 
+  // Since the stateful rebalancer needs to be lazily constructed when the HelixManager instance is
+  // ready, the GenericHelixController is not constructed with a stateful rebalancer. This wrapper
+  // is to avoid the complexity of handling a nullable value in the event handling process.
+  private final StatefulRebalancerRef _rebalancerRef = new StatefulRebalancerRef() {
+    @Override
+    protected StatefulRebalancer createRebalancer(HelixManager helixManager) {
+      return new WagedRebalancer(helixManager);
+    }
```

Comment:
Could we add a TODO here for making it configurable (based on ClusterConfig for example) in the future as we have different versions of stateful rebalancers?

Ideally we won't have to import WagedRebalancer explicitly (which seems like we're creating a dependency between **Generic**HelixController and a specific rebalancer class). I'm fine with creating an issue and getting to it later.

Reply:
The vision is not clear to me. I don't think we have a solid idea for the TODO. For example, relying on the cluster config to set up the rebalancer type may be invalid. If we have multiple optional stateful rebalancers, they will be configured in the ideal state for each resource, instead of cluster config.

More likely, we will need to init all the stateful rebalancers here. It is just too far even for planning. For now, please assume we will only have one stateful rebalancer.

However, making the code generic enough for the future extension is always a good idea. That why I added the interface according to your suggestion : )

Reply:
In addition, about your 2nd paragraph comment, I guess what you concerned about is that the rebalancer class should be only configurable in the IS. So the GenericHelixController should not be aware of the resource level thing. However, please note that even there is no resource using the WAGED rebalancer, the controller will still need to init the stateful rebalancer instance.

I agree this design can be improved. But I would prefer to keep it simple before the rebalancer logic and the algorithm are stable enough.

For now, let me add a TODO.